File: java/src/main/java/util/ParseUtils.java
package util;

import exception.ConfigErrorException;
import model.SensorData;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

public class ParseUtils {

    private static final String PROPERTIES_FILE = "config.properties";
    private static Properties props;

    public static Properties getProperties() {
        if(props == null) {
            ClassLoader loader = Thread.currentThread().getContextClassLoader();
            props = new Properties();
            try(InputStream resourceStream = loader.getResourceAsStream(PROPERTIES_FILE)) {
                props.load(resourceStream);
                if(props.getProperty("reader").isEmpty() || props.getProperty("hci").isEmpty() || props.getProperty("mac").isEmpty()) {
                    throw new ConfigErrorException("Configuration(s) missing in config.properties");
                }
            } catch (IOException e) {
                throw new ConfigErrorException(e);
            }
        }

        return props;

    }

    public static SensorData parseSensorData(String[] bytes) {

        double temperature = Integer.valueOf(bytes[2]+bytes[1],16).shortValue() / 10.0;
        int humidity = Integer.parseInt(bytes[4], 16 );
        int battery = Integer.parseInt(bytes[9], 16 );

        SensorData sensorData = new SensorData();
        sensorData.setTemperature(temperature);
        sensorData.setHumidity(humidity);
        sensorData.setBattery(battery);
        return sensorData;
    }
}

File: javascript/src/main/util/parse-utils.js
var SensorData = require('../model/sensor-data');

function ParseUtils() {};

ParseUtils.parseSensorData = function(bytes) { 
	var temperature = parseInt(bytes.substr(4, 2)+bytes.substr(2, 2), 16);
	if(temperature > 0x8000) {
	    temperature = temperature - 0x10000;
    }
    temperature = temperature / 10.0;
    var humidity = parseInt(bytes.substr(8, 2), 16);
    var battery = parseInt(bytes.substr(18, 2), 16);

    var sensorData = new SensorData();
    sensorData.setTemperature(temperature);
    sensorData.setHumidity(humidity);
    sensorData.setBattery(battery);
    return sensorData;
};

module.exports = ParseUtils;